File: sdks/python/src/opik/integrations/langchain/retrieval_contract.py
from typing import Any, Dict, Mapping, Optional

OPIK_KIND_KEY = "opik.kind"
OPIK_PROVIDER_KEY = "opik.provider"
OPIK_OPERATION_KEY = "opik.operation"

RETRIEVAL_KIND = "retrieval"
DEFAULT_RETRIEVAL_OPERATION = "retrieve"

_GENERIC_RETRIEVER_IDS = {
    "retriever",
    "base",
    "vectorstore",
    "vector_store",
    "langchain",
    "langchain_core",
    "langchain_community",
}


def _normalize_provider(value: Any) -> Optional[str]:
    if not isinstance(value, str):
        return None

    normalized = value.strip().lower().replace(" ", "_")
    if not normalized or normalized in _GENERIC_RETRIEVER_IDS:
        return None

    return normalized


def _infer_provider_from_serialized(run_dict: Dict[str, Any]) -> Optional[str]:
    serialized = run_dict.get("serialized")
    if not isinstance(serialized, Mapping):
        return None

    identifiers = serialized.get("id")
    if not isinstance(identifiers, list):
        return None

    for identifier in reversed(identifiers):
        normalized = _normalize_provider(identifier)
        if normalized:
            return normalized

    return None


def build_retrieval_metadata(
    run_dict: Dict[str, Any], base_metadata: Dict[str, Any]
) -> Dict[str, Any]:
    metadata = dict(base_metadata)

    metadata.setdefault(OPIK_KIND_KEY, RETRIEVAL_KIND)
    metadata.setdefault(OPIK_OPERATION_KEY, DEFAULT_RETRIEVAL_OPERATION)

    provider = _normalize_provider(metadata.get("ls_provider"))
    if provider is None:
        provider = _infer_provider_from_serialized(run_dict)

    if provider is not None:
        metadata.setdefault(OPIK_PROVIDER_KEY, provider)

    return metadata

File: sdks/typescript/src/opik/integrations/opik-langchain/src/retrievalContract.ts
import { Serialized } from "@langchain/core/load/serializable";

type JsonNode = Record<string, unknown>;

const OPIK_KIND_KEY = "opik.kind";
const OPIK_PROVIDER_KEY = "opik.provider";
const OPIK_OPERATION_KEY = "opik.operation";

const RETRIEVAL_KIND = "retrieval";
const DEFAULT_RETRIEVAL_OPERATION = "retrieve";

const GENERIC_RETRIEVER_IDS = new Set([
  "retriever",
  "base",
  "vectorstore",
  "vector_store",
  "langchain",
  "langchain_core",
  "langchain_community",
]);

const normalizeProvider = (value: unknown): string | undefined => {
  if (typeof value !== "string") {
    return undefined;
  }

  const normalized = value.trim().toLowerCase().replaceAll(" ", "_");
  if (!normalized || GENERIC_RETRIEVER_IDS.has(normalized)) {
    return undefined;
  }

  return normalized;
};

const inferProviderFromSerialized = (
  retriever: Serialized
): string | undefined => {
  const identifiers = retriever.id;
  if (!Array.isArray(identifiers)) {
    return undefined;
  }

  for (let i = identifiers.length - 1; i >= 0; i -= 1) {
    const provider = normalizeProvider(identifiers[i]);
    if (provider) {
      return provider;
    }
  }

  return undefined;
};

export const buildRetrieverMetadata = (
  retriever: Serialized,
  metadata?: JsonNode
): JsonNode => {
  const mergedMetadata: JsonNode = { ...(metadata ?? {}) };

  if (!(OPIK_KIND_KEY in mergedMetadata)) {
    mergedMetadata[OPIK_KIND_KEY] = RETRIEVAL_KIND;
  }
  if (!(OPIK_OPERATION_KEY in mergedMetadata)) {
    mergedMetadata[OPIK_OPERATION_KEY] = DEFAULT_RETRIEVAL_OPERATION;
  }

  const provider =
    normalizeProvider(mergedMetadata.ls_provider) ??
    inferProviderFromSerialized(retriever);

  if (provider && !(OPIK_PROVIDER_KEY in mergedMetadata)) {
    mergedMetadata[OPIK_PROVIDER_KEY] = provider;
  }

  return mergedMetadata;
};
